### ISSUE TITLE
Extract features using Yolov8-cls

### ISSUE THREAD

**Issue by Naofel-eal** [ultralytics/ultralytics#6843]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello,

I'm a student currently engaged in developing a face classification system. This system processes video input to identify and group faces of each individual. My current approach involves utilizing Yolov8n for face detection in each video frame. However, I'm encountering a challenge with the face comparison process. I've been using the Deepface library for comparing two faces to determine if they belong to the same person, but I'm finding that Deepface is quite slow for my needs.

Given this, I'm exploring the possibility of training a Yolov8 classifier on a person dataset and then modifying it by removing the softmax layer to use Yolo as a feature extractor. I came across an [issue](https://github.com/ultralytics/ultralytics/issues/3673) where someone implemented a similar approach, but as a beginner, I'm uncertain if this solution is suitable for my project. My primary goal is to extract features from two faces, compare them, and calculate the distance to ascertain if they represent the same individual.

I would greatly appreciate any insights or suggestions on whether this approach is feasible and effective for my requirements.

**Comment by Naofel-eal** [comment]
Thank you very much for your response! Given that the faces in the input videos to the program will be unknown to the model, as it hasn't been trained on them, to obtain the most representative facial features from the model's output, would you recommend using a Yolov8 trained on faces for detection, or a Yolov8-cls trained on people?
And one last question, do you know how it's possible to retrieve the output of the penultimate layer?

**Comment by Naofel-eal** [comment]
Thank you for your feedback @glenn-jocher ! 
I successfully extracted a tensor of size 1280 by commenting out in the [forward](https://github.com/ultralytics/ultralytics/blob/093943e37589f2ee9e9af5761ee116fdad4265bd/ultralytics/nn/modules/head.py#L164) method of the [Classify](https://github.com/ultralytics/ultralytics/blob/093943e37589f2ee9e9af5761ee116fdad4265bd/ultralytics/nn/modules/head.py#L150) class in the [head.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/nn/modules/head.py) file, the last two layers: the softmax layer and the linear layer:
```python
def forward(self, x):
        """Performs a forward pass of the YOLO model on input image data."""
        if isinstance(x, list):
            x = torch.cat(x, 1)
        x = self.conv(x)
        x = self.pool(x)
        x = x.flatten(1)
        x = self.drop(x)
        # x = self.linear(x)
        # x = x.softmax(1)
        return x
```
I also modified the [postprocess](https://github.com/ultralytics/ultralytics/blob/093943e37589f2ee9e9af5761ee116fdad4265bd/ultralytics/models/yolo/classify/predict.py#L40) method of [ClassificationPredictor](https://github.com/ultralytics/ultralytics/blob/093943e37589f2ee9e9af5761ee116fdad4265bd/ultralytics/models/yolo/classify/predict.py#L10) in the [predict.py](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/classify/predict.py) file to return the feature tensor directly without a Result object:
```python
def postprocess(self, preds, img, orig_imgs):
        """Post-processes predictions to return Results objects."""
        if not isinstance(orig_imgs, list):  # input images are a torch.Tensor, not a list
            orig_imgs = ops.convert_torch2numpy_batch(orig_imgs)

        results = []
        return preds
        # for i, pred in enumerate(preds):
        #     orig_img = orig_imgs[i]
        #     img_path = self.batch[0][i]
        #     results.append(Results(orig_img, path=img_path, names=self.model.names, probs=pred))
        # return results
```

However, when I calculate the cosine distance between two tensors, I get very poor results that do not allow me to determine if the two compared faces belong to the same person.
I trained my Yolov8n-cls on the [Celebrity Face Image Dataset](https://www.kaggle.com/datasets/vishesh1412/celebrity-face-image-dataset), which contains about a hundred images for each of the 17 different people.

Following these poor results, I therefore have 2 questions:
- Have I retrieved the output from the right layer to get the features of the face sent to the model?
- Would you recommend me to retrain Yolov8n-cls on a larger dataset in order to obtain more descriptive features of the faces?

Thank you immensely for the help you are providing me 🙏!

**Comment by glenn-jocher** [comment]
@Naofel-eal,

1. Modifying the `forward` method as you did is indeed a common approach to retrieve features from a neural network. Your edit to extract the 1280-sized tensor should be outputting the activations from the penultimate layer, which are supposed to be rich in features.

2. Training on a larger and more diverse dataset is generally advisable to capture more generalizable features. The Celebrity Face Image Dataset limited to 17 individuals may not provide enough variability. Consider a dataset with broader intra-class variety.

Keep up the good work, and remember that the effectiveness of feature vectors greatly depends on the representativeness of your training data and the network's ability to generalize. 🌟

**Comment by manishp54** [comment]
@glenn-jocher @Naofel-eal, I have downloaded the repository of YOLOv8  and have updated the code same as they have done. now tell me how i can run that code?   i want to store the features of image but i don't know how to run the file from repository. i'm using following code but i've got nothing.

**Comment by glenn-jocher** [comment]
@manishp54, to use your modified YOLOv8 for feature extraction, you would run it similarly to how you'd run any inference, just bear in mind that you need to ensure your `predict.py` script is capable of handling the modified model output.

Using your provided command seems mostly correct, assuming `predict.py` is set up to accept command line arguments and process the model output. Ensure your modified `predict.py` returns the feature tensor as you expect.

If problems persist, verify the script modifications reflect your needs, and check for any errors in the console output. Additionally, ensure the correct file paths and verify if the model weights are properly loaded.

Remember to consult the Ultralytics Docs for guidance on running inference.

**Comment by Naofel-eal** [comment]
Hi @glenn-jocher, thank you very much for your help. 
So, we tried using the 1280-size vector from Yolov8-cls output to see if two faces belong to the same person, but it turned out that this vector was not feature-rich enough for an effective face comparison. Therefore, we turned to VGG-Face, extracting a 2622-size vector that was rich enough to allow us to perform an effective comparison.

**Comment by glenn-jocher** [comment]
@Naofel-eal, glad to hear you've found a solution that works for your project. It's not uncommon for different architectures to yield varying results in feature richness for specific tasks. VGG-Face is indeed well-tuned for facial recognition tasks. Best of luck with your continued work! 🌟
